lib/stitch_plan/lock_stitch.py

```diff
@@ -0,0 +1,202 @@
+import re
+from copy import copy
+from math import degrees
+
+from inkex import DirectedLineSegment, Path
+from shapely.geometry import LineString
+from shapely.ops import substring
+
+from ..i18n import _
+from ..svg import PIXELS_PER_MM
+from ..utils import string_to_floats
+from .stitch import Stitch
+
+
+class LockType:
+    def __init__(self, lock_id=None, name=None, path=None, path_type=None,
+                 scale_percent=100, scale_absolute=0.7):
+
+        self.id: str = lock_id
+        self.name: str = name
+        self.path: str = path
+        self.path_type: str = path_type
+        self.scale_percent: float = scale_percent
+        self.scale_absolute: float = scale_absolute
+
+    def __repr__(self):
+        return "LockType(%s, %s, %s, %s, %s, %s)" % (self.id, self.name, self.path, self.path_type,
+                                                     self.scale_percent, self.scale_absolute)
+
+    def __str__(self):
+        return str(self.id) or ""
+
+    def copy(self, scale_percent=None, scale_absolute=None):
+        cp = copy(self)
+        cp.set('scale_percent', scale_percent or self.scale_percent)
+        cp.set('scale_absolute', scale_absolute or self.scale_absolute)
+        return cp
+
+    def set(self, attribute, value):
+        setattr(self, attribute, value)
+
+    def _check_custom(self):
+        if not self.id == "custom":
+            return
+        if not self.path:
+            self._fallback()
+            return
+
+        if not re.match("^ *[0-9 .,-]*$", self.path):
+            path = Path(self.path)
+            if not path or len(list(path.end_points)) < 3:
+                self._fallback()
+            else:
+                self.set('path_type', "svg")
+        else:
+            path = string_to_floats(self.path, " ")
+            if path:
+                self.set('path_type', "absolute")
+            else:
+                self._fallback()
+
+    def _fallback(self):
+        self.set('path', half_stitch.path)
+        self.set('path_type', half_stitch.path_type)
+
+    def stitches(self, group_stitches, pos):
+        self._check_custom()
+
+        if pos == "end":
+            group_stitches = list(reversed(group_stitches))
+
+        if self.path_type == "absolute":
+            lock_stitches = self._absolute_tie(group_stitches, pos)
+        elif self.path_type == "svg":
+            lock_stitches = self._svg_tie(group_stitches, Path(self.path), pos)
+        else:
+            lock_stitches = self._relative_tie(group_stitches, pos)
```

Comment:
This is probably a good use for class inheritance.  Each type could be its own class (like AbsoluteLock, RelativeLock, and SVGLock).  Those would all inherit from a common class Lock for shared functionality.  Then each one could have its own definition for `stitches()` that does the right thing for that kind of lock.  Custom could be a subclass of SVGLock.

It's late and I think I may not be explaining this well -- I'm happy to answer questions or explain better tomorrow if this doesn't make sense.

Reply:
Subclasses added with an additionally a CustomLock class.